src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+    `__new__`.
+
+    NOTE: Due to https://github.com/python/mypy/issues/5146 subclasses unfortunately need to provide
+    overloaded `__new__` methods, as subclasses do not inherit overloaded function's annotations.
+    Use one of the existing subclasses as a guide on how to do this.
+    """
+
+    flag_names: tuple[str, ...]
+    kwargs: Dict
+    option_type: Any  # NB: This should be some kind of callable that returns PropType
+
+    # NB: Due to https://github.com/python/mypy/issues/5146, we try to keep the parameter list as
+    # short as possible to avoid having to repeat the param in each of the 3 `__new__` specs that
+    # each subclass must provide.
+    # If you need additional information, this class follows the "Builder" pattern. Use the
+    # "builder" methods to attach additional information. (E.g. see `advanced` method)
+    # NB: We define `__new__` rather than `__init__` because otherwise subclasses would have to
+    # define 3 `__new__`s AND `__init__`.
+    def __new__(
+        cls,
+        *flag_names: str,
+        default: Any,
+        option_type: Type | None = None,
+        help: str,
+    ):
+        self = super().__new__(cls)
+        self.flag_names = flag_names
+        if option_type is None:
+            option_type = cls.option_type
+
+        self.kwargs = dict(
+            type=option_type,
+            default=default,
+            help=help,
+        )
+        return self
+
+    @overload
+    def __get__(self, obj: None, objtype: Any) -> OptionBase[PropType]:
+        ...
+
+    @overload
+    def __get__(self, obj: object, objtype: Any) -> PropType:
+        ...
+
+    def __get__(self, obj, objtype):
+        if obj is None:
+            return self
+        long_name = self.flag_names[-1]
+        option_value = getattr(obj.options, long_name[2:].replace("-", "_"))
+        if option_value is None:
+            return None
+        return self._convert_(option_value)
+
+    # Subclasses can override if necessary
+    def _convert_(self, val: Any) -> PropType:
+        return cast("PropType", self.kwargs["type"](val))
+
+    # ===== "Builder" Methods =====
+    def advanced(self) -> OptionBase[PropType]:
+        self.kwargs["advanced"] = True
+        return self
+
+    def from_file(self) -> OptionBase[PropType]:
+        self.kwargs["fromfile"] = True
+        return self
+
+    def metavar(self, metavar: str) -> OptionBase[PropType]:
+        self.kwargs["metavar"] = metavar
+        return self
+
+    def mutually_exclusive_group(self, mutually_exclusive_group: str) -> OptionBase[PropType]:
+        self.kwargs["mutually_exclusive_group"] = mutually_exclusive_group
+        return self
+
+    def default_help_repr(self, default_help_repr: str) -> OptionBase[PropType]:
+        self.kwargs["default_help_repr"] = default_help_repr
+        return self
+
+    def will_be_removed(self, *, version: str, hint: str) -> OptionBase[PropType]:
```

Comment:
How about `deprecated()`? We use that phrase much more.

Reply:
FWIW I chose `removed` so we wouldn't have to prefix the param names. Otherwise the "version" is ambiguous (is it the version it was deprecated in?).

What do you think of `deprecated(*, removal_version, hint)`? Seems OK?

Reply:
Yes, I love that suggestion! And it fits with how `base.deprecated` does everything :)